### PR TITLE
JDK-8300857: State return value for Types.asElement(NoType) explicitly

### PR DESCRIPTION
**Pull request by jddarcy**
Just sending out the proposed API changes for now, will add tests later.

Please also review the corresponding CSR [JDK-8300951](https://bugs.openjdk.org/browse/JDK-8300951).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change requires CSR request [JDK-8300951](https://bugs.openjdk.org/browse/JDK-8300951) to be approved
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8300857](https://bugs.openjdk.org/browse/JDK-8300857): State return value for Types.asElement(NoType) explicitly
 * [JDK-8300951](https://bugs.openjdk.org/browse/JDK-8300951): State return value for Types.asElement(NoType) explicitly (**CSR**)

### Reviewers
 * [Jonathan Gibbons](https://openjdk.org/census#jjg) (@jonathan-gibbons - **Reviewer**) ⚠️ Review applies to [a60301ce](https://git.openjdk.org/jdk/pull/12159/files/a60301cecf4c0add9b2ebc3ecb2ac409bfe056f9)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12159/head:pull/12159` \
`$ git checkout pull/12159`

Update a local copy of the PR: \
`$ git checkout pull/12159` \
`$ git pull https://git.openjdk.org/jdk pull/12159/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12159`

View PR using the GUI difftool: \
`$ git pr show -t 12159`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12159.diff">https://git.openjdk.org/jdk/pull/12159.diff</a>

</details>

### CODE REVIEW

File: src/java.compiler/share/classes/javax/lang/model/util/Types.java [hunk 1]
```diff
@@ -49,6 +49,19 @@ public interface Types {
      * Returns {@code null} if the type is not one with a
      * corresponding element.
      *
+     * <p>Types without corresponding elements include:
+     * <ul>
+     * <li>{@linkplain TypeKind#isPrimitive() primitive types}
+     * <li>{@linkplain TypeKind#EXECUTABLE executable types}
+     * <li>{@linkplain TypeKind#MODULE module pseudo-types}
+     * <li>{@linkplain TypeKind#NONE "none" pseudo-types}
+     * <li>{@linkplain TypeKind#NULL  null types}
```
Comment (pavelrappo) [inline src/java.compiler/share/classes/javax/lang/model/util/Types.java:58]: Was that extra whitespace between NULL and null intended?

**Review by pavelrappo — COMMENTED** [review]
Thanks for doing this. It looks good. However, I'll leave it to compiler experts to formally approve this PR.

**Comment by jddarcy** [comment]
PS I purposely did not provide an explicit specification on how a TypeMirror with a kind of OTHER or ERROR might be handled here to preserve implementation flexibility.

As an aside, if we had more time for the original JSR 269 effort, we would likely have worked so that there was a PrimitiveElement enum to host the element form of the built-in primitive types, but we ran out of time do to that work. That would have allowed a fuller round-trip relation of

element.asType().asElement()

returning something equal to the initial value.

**Comment by jddarcy** [comment]
Quick discussion of the revised version of the patch:

* After looking at the javac implementation, I added spec and implementation support for the type mirrors of packages and modules returning the element for the package or module. This was a very small addendum to add support and allows round-tripping in more cases.

* Neither union types nor intersection types are in the explicitly mapped or explicitly not mapped lists. This types are generally returned from trees and elements modeling the inside of method bodies rather than the declaration-level structures in the standardized API.

The remaining explicitly listed types-without-elements are straightforward to generate via the standard APIs.

File: src/java.compiler/share/classes/javax/lang/model/util/Types.java [hunk 2]
```diff
@@ -45,10 +45,22 @@ public interface Types {
 
     /**
      * Returns the element corresponding to a type.
-     * The type may be a {@code DeclaredType} or {@code TypeVariable}.
+     * The type may be a {@link DeclaredType} or {@link TypeVariable},
+     * or a pseudo-type for a {@linkplain TypeKind#PACKAGE package} or
+     * {@linkplain TypeKind#MODULE module}.
```
Comment (jonathan-gibbons) [inline src/java.compiler/share/classes/javax/lang/model/util/Types.java:50]: Commas look erratic.

`The type may be A or B, or C`

File: test/langtools/tools/javac/processing/model/util/types/TestAsElement.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        expectNullAsElement(typeUtils.getWildcardType(objectElement.asType(), null));
+
+        // Loop over the ExecutableTypes for Object's methods
+        for(var methodElt : ElementFilter.methodsIn(objectElement.getEnclosedElements())) {
+            expectNullAsElement(methodElt.asType());
+        }
+    }
+
+
+
```
Comment (jonathan-gibbons) [inline test/langtools/tools/javac/processing/model/util/types/TestAsElement.java:78]: excess blank lines?

File: test/langtools/tools/javac/processing/model/util/types/TestAsElement.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    private void expectNullAsElement(TypeMirror typeMirror) {
+        var e = typeUtils.asElement(typeMirror);
+        if (e != null) {
+            throw new RuntimeException("Unexpected non-null value " + e);
+        }
+    }
+
+    private void testRoundTripCases() {
+        expectRoundTrip(eltUtils.getPackageElement("java.lang"));
```
Comment (jonathan-gibbons) [inline test/langtools/tools/javac/processing/model/util/types/TestAsElement.java:87]: Maybe add a type element as well as the package element and module element

**Review by jonathan-gibbons — APPROVED** [review]
approved with quibbles for optional update

**Comment by jonathan-gibbons** [comment]
I like the presence of the new test ;-)

**Pull request closed without merging**